src/Aspire.Hosting.Keycloak/KeycloakResouceBuilderExtensions.cs

```diff
@@ -0,0 +1,72 @@
+    private const string AdminEnvVarName = "KEYCLOAK_ADMIN";
+    private const string AdminPasswordEnvVarName = "KEYCLOAK_ADMIN_PASSWORD";
+    private const int DefaultContainerPort = 8080;
+
+    /// <summary>
+    /// Adds a Keycloak container to the application model.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <param name="name">The name of the resource. </param>
+    /// <param name="port">The host port that the underlying container is bound to when running locally.</param>
+    /// <param name="admin">The parameter used as the admin for the Keycloak resource. If <see langword="null"/> a default value will be used.</param>
+    /// <param name="adminPassword">The parameter used as the admin password for the Keycloak resource. If <see langword="null"/> a default password will be used.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<KeycloakResource> AddKeycloak(
+        this IDistributedApplicationBuilder builder,
+        string name,
+        int? port = null,
+        IResourceBuilder<ParameterResource>? admin = null,
+        IResourceBuilder<ParameterResource>? adminPassword = null)
+    {
+        var passwordParameter = adminPassword?.Resource ?? ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-password");
+
+        var resource = new KeycloakResource(name, admin?.Resource, passwordParameter);
+
+        var keycloak = builder
+            .AddResource(resource)
+            .WithImage(KeycloakContainerImageTags.Image)
+            .WithImageRegistry(KeycloakContainerImageTags.Registry)
+            .WithImageTag(KeycloakContainerImageTags.Tag)
+            .WithHttpEndpoint(port: port, targetPort: DefaultContainerPort)
+            .WithEnvironment(context =>
+            {
+                context.EnvironmentVariables[AdminEnvVarName] = resource.AdminReference;
+                context.EnvironmentVariables[AdminPasswordEnvVarName] = resource.AdminPasswordParameter;
+            });
+
+        if (builder.ExecutionContext.IsRunMode)
+        {
+            keycloak.WithArgs("start-dev");
+        }
+        else
+        {
+            keycloak.WithArgs("start");
+        }
+
+        return keycloak;
+    }
+
+    /// <summary>
+    /// Adds a named volume for the data folder to a Keycloak container resource.
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the volume. Defaults to an auto-generated name based on the application and resource names.</param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only volume.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<KeycloakResource> WithDataVolume(this IResourceBuilder<KeycloakResource> builder, string? name = null, bool isReadOnly = false)
```

Comment:
Do you have any details on what exactly will get persisted in the volume mapped to `/opt/keycloak/data`? The only mention of it in the [Keycloak container docs](https://www.keycloak.org/server/containers) I can find is in reference to importing realms by mounting files under the `/opt/keycloak/data/import` directory, but it's not clear to me if there's other scenarios that use the data directory. Will this 

Assuming mounting to `/opt/keycloak/data` is enough to store all data between instances, we should add a `WithDataBindMount` extension method too, similar to the other resources with these methods.

Reply:
From what I have seen so far, everything I've used gets persisted to the mounted volume, including the realm, clients, client scopes, users and roles.